### PR TITLE
move sui-oracle move package to public repo

### PR DESCRIPTION
**Pull request by patrickkuo**
## Description 

Sui Oracle is a general purpose oracle that can be used to fetch data from external sources and feed it to the Sui Chain.
It is designed to be modular and extensible, so that it can be easily extended to support new data sources and new data types.

## Test Plan 

Move test + manual integration tests

## TODOs
We need to add more documentation to explain the design of this package and turn it into an example.

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] protocol change
- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/sui-oracle/move/oracle/sources/simple_oracle.move [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        identifier: String,
+    }
+
+    public fun get_historical_data<K: copy + drop + store, V: store + copy>(
+        oracle: &SimpleOracle,
+        ticker: String,
+        archival_key: K
+    ): Option<Data<V>> {
+        string::append(&mut string::utf8(b"[historical] "), ticker);
+        let historical_data: &Table<K, StoredData<V>> = df::borrow(&oracle.id, ticker);
```
Comment (gdanezis) [inline crates/sui-oracle/move/oracle/sources/simple_oracle.move:47]: Should we be checking here if the historical reading exists?

Comment (gdanezis) [reply crates/sui-oracle/move/oracle/sources/simple_oracle.move:47]: Also adding "[historical] " is rather large, is that a problem?

File: crates/sui-oracle/move/oracle/sources/simple_oracle.move [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+            return option::none()
+        };
+        let data: &StoredData<T> = df::borrow(&oracle.id, ticker);
+        let StoredData { value, sequence_number, timestamp, identifier } = *data;
+        option::some(data::new(value, ticker, sequence_number, timestamp, oracle.address, identifier))
+    }
+
+    /// Create a new shared SimpleOracle object for publishing data.
+    public entry fun create(name: String, url: String, description: String, ctx: &mut TxContext) {
+        let oracle = SimpleOracle { id: object::new(ctx), address: tx_context::sender(ctx), name, description, url };
```
Comment (gdanezis) [inline crates/sui-oracle/move/oracle/sources/simple_oracle.move:66]: It seems here that we only authenticate by address, should we provide a mode where we authenticate by a capability? This allows oracles more flexible security policies when updating.

File: crates/sui-oracle/move/oracle/sources/meta_oracle.move [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+
+    #[test_only]
+    use oracle::decimal_value;
+
+    const EValidDataSizeLessThanThreshold: u64 = 0;
+    const EUnsupportedDataType: u64 = 1;
+
+    struct MetaOracle<T> {
+        oracle_datas: vector<Option<Data<T>>>,
+        threshole: u64,
```
Comment (gdanezis) [inline crates/sui-oracle/move/oracle/sources/meta_oracle.move:24]: typo: "threshold"

File: crates/sui-oracle/move/oracle/sources/meta_oracle.move [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+        time_window_ms: u64,
+        ticker: String,
+        max_timestamp: u64,
+    }
+
+    public fun new<T: copy + drop>(threshole: u64, time_window_ms: u64, ticker: String): MetaOracle<T> {
+        MetaOracle {
+            oracle_datas: vector::empty(),
```
Comment (gdanezis) [inline crates/sui-oracle/move/oracle/sources/meta_oracle.move:32]: gr: "data" is already plural :)

File: crates/sui-oracle/move/oracle/sources/meta_oracle.move [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+
+    struct MetaOracle<T> {
+        oracle_datas: vector<Option<Data<T>>>,
+        threshole: u64,
+        time_window_ms: u64,
+        ticker: String,
+        max_timestamp: u64,
+    }
+
+    public fun new<T: copy + drop>(threshole: u64, time_window_ms: u64, ticker: String): MetaOracle<T> {
```
Comment (gdanezis) [inline crates/sui-oracle/move/oracle/sources/meta_oracle.move:30]: type: "threshold"

File: crates/sui-oracle/move/oracle/sources/meta_oracle.move [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+        };
+        vector::push_back(&mut meta_oracle.oracle_datas, oracle_data);
+    }
+
+    struct TrustedData<T> has copy, drop {
+        value: T,
+        oracles: vector<address>,
+    }
+
+    fun combind<T: copy + drop>(meta_oracle: MetaOracle<T>, ): (vector<T>, vector<address>) {
```
Comment (gdanezis) [inline crates/sui-oracle/move/oracle/sources/meta_oracle.move:53]: "combine"

File: crates/sui-oracle/move/oracle/sources/meta_oracle.move [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+            };
+        };
+        assert!(vector::length(&values) >= threshole, EValidDataSizeLessThanThreshold);
+        (values, oracles)
+    }
+
+    /// take the median value
+    public fun median<T: copy + drop>(meta_oracle: MetaOracle<T>): TrustedData<T> {
+        let (values, oracles) = combind(meta_oracle);
+        let sortedData = quick_sort(values);
```
Comment (gdanezis) [inline crates/sui-oracle/move/oracle/sources/meta_oracle.move:75]: Probably worth having a generic quicksort function for others to use? Not sure how we can do that in move.

**Review by gdanezis — COMMENTED** [review]
Hey Patrick, have a look at the items below.

File: crates/sui-oracle/move/oracle/sources/meta_oracle.move (hunk 7 above)
Comment (patrickkuo) [reply crates/sui-oracle/move/oracle/sources/meta_oracle.move:75]: I try to make this function generic by using bcs (not sure if that's the most efficient way, cc @damirka for comment), it now support all numeric types
and `DecimalValue`. I don't think there are easy way to make it support arbitrary types.
I will make it public for now so other can use it, my plan is to put this in a maths lib so more people can use it.

**Comment by longbowlu** [comment]
Move it to an `example` folder is a good idea. All remaining comments do not seem blocking to me.

**Review by longbowlu — APPROVED** [review]

**Pull request merged**